### PR TITLE
Improve the Pass system

### LINKED ISSUE #1641: Implement `latePass` annotation
The ordering system should be extended: We would like to have a `latePass` (or similar) annotation in addition to the `lastPass` annotation. Annotated passes should be executed as late as possible, without breaking the other ordering constraints.

### PR DESCRIPTION
**Pull request by maximiliankaul**
This PR cleans up the pass ordering system. It removes some logic bugs and tries to keep a strict separation between files (TranslationConfiguration does not do the ordering logic, ...)
New features:
- `ExecuteLate` annotation.
- `ExecuteBefore` now has `soft`/`hard` dependencies.
- `Pass`es (their `KClass`) can now be queried for `isFirstPass`, `softDependencies`, ...

Fixes #1641 

Todo;
- [x] Clean up
- [x] Mermaid @oxisto do we still want this, or is the current output ok?
- [x] move first / last out of `workingList`
- [x] Think about unjustifiable dependencies (soft dependency not registered)
- [x] implement `soft` / `hard` for `ExecuteBefore` as requested by @KuechA

### CODE REVIEW

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt [hunk 1]
```diff
@@ -651,171 +651,13 @@ private constructor(
...
-                }
-            }
-            return workingList
-        }
-
-        /**
-         * Builds a markdown representation of a pass dependency graph, based on
-         * [Mermaid](https://mermaid.js.org) syntax.
-         */
-        private fun buildMermaid(
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt:728]: Personally, I found the mermaid output quite useful on multiple occasions. Is there a specific reason to get rid of it?

Comment (maximiliankaul) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt:728]: Yes, I don't know how to adapt it to he current code with parallel passes... I'm happy to receive a commit. Probalby at the end of `order()`

Comment (maximiliankaul) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt:728]: It's back [bf706dc](https://github.com/Fraunhofer-AISEC/cpg/pull/1645/commits/bf706dc4bde2235751c7d0ce456627955bd11082)

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt [hunk 2]
```diff
@@ -651,171 +651,73 @@ private constructor(
...
-
         /**
          * Builds a markdown representation of a pass dependency graph, based on
          * [Mermaid](https://mermaid.js.org) syntax.
          */
-        private fun buildMermaid(
-            softDependencies: MutableMap<KClass<out Pass<*>>, MutableSet<KClass<out Pass<*>>>>,
-            hardDependencies: MutableMap<KClass<out Pass<*>>, MutableSet<KClass<out Pass<*>>>>
-        ): String {
+        private fun buildMermaid(passes: List<KClass<out Pass<out Node>>>): String {
```
Comment (maximiliankaul) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt:665]: This became quiet a bit of code. Should we move this to a separate file?

Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt:665]: Yes that would make sense

Comment (maximiliankaul) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationConfiguration.kt]: [faef9e2](https://github.com/Fraunhofer-AISEC/cpg/pull/1645/commits/faef9e2d66970f125800e282780ae94226e22e12)

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+/** Helper to create a mermaid graph given a list of passes. */
+object MermaidHelper {
+    /** Helper function to replace the first and last passes names by their identifier. */
+    private const val UNKNOWNPASS = "UnknownPass"
+    private const val FIRSTPASSESSUBGRAPHIDENTIFIER = "FirstPassesSubgraph"
+    private const val NORMALPASSESUBGRAPHIDENTIFIER = "NormalPassesSubgraph"
+    private const val LASTPASSESSUBGRAPHIDENTIFIER = "LastPassesSubgraph"
+    private const val FIRSTPASSIDENTIFIER = "FirstPass"
+    private const val LASTPASSIDEINTIFIER = "LastPass"
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt:46]: typo

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import de.fraunhofer.aisec.cpg.passes.hardExecuteBefore
+import de.fraunhofer.aisec.cpg.passes.isFirstPass
+import de.fraunhofer.aisec.cpg.passes.isLastPass
+import de.fraunhofer.aisec.cpg.passes.softDependencies
+import de.fraunhofer.aisec.cpg.passes.softExecuteBefore
+import kotlin.reflect.KClass
+
+/** Helper to create a mermaid graph given a list of passes. */
+object MermaidHelper {
+    /** Helper function to replace the first and last passes names by their identifier. */
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt:40]: this doc string is probably part of line 48

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import de.fraunhofer.aisec.cpg.passes.hardDependencies
+import de.fraunhofer.aisec.cpg.passes.hardExecuteBefore
+import de.fraunhofer.aisec.cpg.passes.isFirstPass
+import de.fraunhofer.aisec.cpg.passes.isLastPass
+import de.fraunhofer.aisec.cpg.passes.softDependencies
+import de.fraunhofer.aisec.cpg.passes.softExecuteBefore
+import kotlin.reflect.KClass
+
+/** Helper to create a mermaid graph given a list of passes. */
+object MermaidHelper {
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt:39]: Kotlin supports top-level functions, there would not be a need for a helper object, you can just directly add the two functions to the file here.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    /** Helper function to replace the first and last passes names by their identifier. */
+    private const val UNKNOWNPASS = "UnknownPass"
+    private const val FIRSTPASSESSUBGRAPHIDENTIFIER = "FirstPassesSubgraph"
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/MermaidHelper.kt:42]: the canonical naming would probably be `FIRST_PASSES_SUBGRAPH_IDENTIFIER` or something similar.

**Review by oxisto — COMMENTED** [review]
only minor formatting issues.

**Review by oxisto — APPROVED** [review]
LGTM except the minor formatting issues int he mermaid file that should be resolved first.

**Pull request merged**